src/models/dirac_mixture_multi_publisher_model.py

```diff
@@ -0,0 +1,632 @@
+        baseline_impression_vector = []
+        for i in range(p):
+            if not i in single_pub_points.keys():
+                raise AssertionError(f"Cannot find single pub reach point for pub {i}")
+            rp = single_pub_points[i][0]
+            prior_marginal_frequency_histograms.append(rp.zero_included_histogram)
+            baseline_impression_vector.append(rp.impressions[i])
+        return (
+            np.array(prior_marginal_frequency_histograms),
+            np.array(baseline_impression_vector),
+        )
+
+    @staticmethod
+    def obtain_observable_directions(
+        reach_points: List[ReachPoint], baseline_impression_vector: np.ndarray
+    ) -> np.ndarray:
+        """Obtain observable_directions as an input of MultivariateMixedPoissonOptimizer.
+
+        As mentioned in the description of , the observable_directions are obtain
+        by standardizing the impression vector of each training point with the
+        baseline_impression_vector.
+        """
+        return (
+            np.array([list(rp.impressions) for rp in reach_points])
+            / baseline_impression_vector[np.newaxis, :]
+        )
+
+    @classmethod
+    def obtain_frequency_histograms_on_observable_directions(
+        cls, reach_points: List[ReachPoint]
+    ) -> np.ndarray:
+        """Obtain frequency_histograms_on_observable_directions as an input of MultivariateMixedPoissonOptimizer."""
+        return np.array([rp.zero_included_histogram for rp in reach_points])
+
+    def _fit(self):
+        if self._fit_computed:
+            return
+        prior_hists, self.baseline_imps = self.obtain_marginal_frequency_histograms(
+            reach_points=self._data
+        )
+        obs_dirs = self.obtain_observable_directions(
+            reach_points=self._data, baseline_impression_vector=self.baseline_imps
+        )
+        hists_on_obs_dirs = self.obtain_frequency_histograms_on_observable_directions(
+            reach_points=self._data
+        )
+        while self.ncomponents > 0:
+            self.optimizer = MultivariateMixedPoissonOptimizer(
+                observable_directions=obs_dirs,
+                frequency_histograms_on_observable_directions=hists_on_obs_dirs,
+                prior_marginal_frequency_histograms=prior_hists,
+                ncomponents=self.ncomponents,
+                rng=self.rng,
+            )
+            try:
+                self.optimizer.fit()
+                break
+            except Exception as inst:
+                # There is a tiny chance of exception when cvxpy mistakenly
+                # thinks the problem is non-convex due to numerical errors.
+                # If this occurs, it is likely that we have a large number
+                # of components.  In this case, try reducing the number of
+                # components.
+                logging.vlog(1, f"Optimizer failure: {inst}")
+                self.ncomponents = int(self.ncomponents / 2)
```

Comment:
int(self.ncomponents / 2) is equivalent to self.ncomponents // 2 when of course self.ncomponents is int.

Reply:
Thanks, changed.